plonky2x/core/src/frontend/ecc/curve25519/ed25519/eddsa.rs

```diff
@@ -0,0 +1,281 @@
+    }
+
+    fn test_curta_eddsa_verify_sigs(variable_msg_len: bool) {
+        utils::setup_logger();
+
+        const MAX_MSG_LEN_BYTES: usize = 192;
+        const NUM_SIGS: usize = 3;
+        type FF = Ed25519ScalarField;
+        let mut builder = DefaultBuilder::new();
+
+        let pkeys = builder.read::<ArrayVariable<CompressedEdwardsYVariable, NUM_SIGS>>();
+        let signatures = builder.read::<ArrayVariable<EDDSASignatureVariable<FF>, NUM_SIGS>>();
+        let messages = builder.read::<ArrayVariable<BytesVariable<MAX_MSG_LEN_BYTES>, NUM_SIGS>>();
+        if variable_msg_len {
+            let message_lens = builder.read::<ArrayVariable<U32Variable, NUM_SIGS>>();
+            builder.curta_eddsa_verify_sigs(messages, Some(message_lens), signatures, pkeys);
+        } else {
+            builder.curta_eddsa_verify_sigs(messages, None, signatures, pkeys);
+        }
+
+        let circuit = builder.build();
+
+        // Generate random messages and private keys
+        let mut test_messages: Vec<[u8; MAX_MSG_LEN_BYTES]> = Vec::new();
+        let mut test_message_lens = Vec::new();
+        let mut test_pub_keys = Vec::new();
+        let mut test_signatures = Vec::new();
+
+        let mut csprng = OsRng;
+        for _i in 0..NUM_SIGS {
+            // Generate random length
+            let msg_len: u32;
+            if variable_msg_len {
+                msg_len = rand::thread_rng().gen_range(1..MAX_MSG_LEN_BYTES) as u32;
+                test_message_lens.push(msg_len);
+            } else {
+                msg_len = MAX_MSG_LEN_BYTES as u32;
+            }
+            let mut test_message = Vec::new();
+            for _ in 0..msg_len {
+                test_message.push(rand::thread_rng().gen_range(0..255));
+            }
+
+            let test_signing_key = SigningKey::generate(&mut csprng);
+            let test_pub_key = test_signing_key.verifying_key();
+            let test_signature = test_signing_key.sign(&test_message);
+
+            test_message.resize(MAX_MSG_LEN_BYTES, 0);
+            test_messages.push(test_message.try_into().unwrap());
+            test_pub_keys.push(CompressedEdwardsY(test_pub_key.to_bytes()));
+            test_signatures.push(EDDSASignatureVariableValue {
+                r: CompressedEdwardsY(*test_signature.r_bytes()),
+                s: BigUint::from_bytes_le(test_signature.s_bytes()),
+            });
+        }
+
+        let mut input = circuit.input();
+        input.write::<ArrayVariable<CompressedEdwardsYVariable, NUM_SIGS>>(test_pub_keys.clone());
+        input.write::<ArrayVariable<EDDSASignatureVariable<FF>, NUM_SIGS>>(test_signatures.clone());
+        input.write::<ArrayVariable<BytesVariable<MAX_MSG_LEN_BYTES>, NUM_SIGS>>(
+            test_messages.clone(),
+        );
+        if variable_msg_len {
+            input.write::<ArrayVariable<U32Variable, NUM_SIGS>>(test_message_lens.clone());
+        }
+
+        let (proof, output) = circuit.prove(&input);
```

Comment:
Should we also have a negative test here that should panic if the signature is invalid?

Reply:
I added a failed signature test.